### PR TITLE
Machine options and support v5 `guard` and `actors`

### PR DESCRIPTION
**Pull request by farskid**
- Support both `actors` and `guards` in machine implementations and if both are present, merges them into a single object
- Support v5 `guard` keyword for transition conditions and choose action conditions. If both `cond` and `guard` are present on a transaction, picks `guard`.

### CODE REVIEW

File: packages/machine-extractor/src/__tests__/options.test.ts [hunk 1]
```diff
@@ -21,4 +21,58 @@ describe('Options', () => {
 
     expect(t.isObjectMethod(node)).toBeTruthy();
   });
+
+  it.only('Should extract implementations provided in machine options', () => {
```
Comment (davidkpiano) [inline packages/machine-extractor/src/__tests__/options.test.ts:25]:
```suggestion
  it('Should extract implementations provided in machine options', () => {
```

File: packages/machine-extractor/src/extractMachinesFromFile.ts [hunk 2]
```diff
@@ -13,15 +13,17 @@ export const extractMachinesFromFile = (
...
   return {
-    machines: machineNodes.map((node) =>
-      getMachineExtractResult({
-        file,
-        fileContent,
-        node,
-      }),
-    ),
+    machines,
+    options: machines.map((m) => m?.getAllMachineOptions()),
```
Comment (Andarist) [inline packages/machine-extractor/src/extractMachinesFromFile.ts:26]: Why do we need this here? Can't we just always get *specific* options from a *specific* machine when needed?

File: packages/machine-extractor/src/MachineExtractResult.ts [hunk 3]
```diff
@@ -708,6 +709,46 @@ export class MachineExtractResult {
     return delays.toObject();
   };
 
+  getAllMachineOptions = () => {
```
Comment (Andarist) [inline packages/machine-extractor/src/MachineExtractResult.ts:712]: we are not really super strict about it today but I think that currently we call those `implementations` (and not `options`)

**Review by davidkpiano — APPROVED** [review]

**Comment by Andarist** [comment]
> Support both actors and guards in machine implementations and if both are present, merges them into a single object

This sounds incorrect. `guards` and `actors` probably shouldn't be merged into a single object. Or have you maybe meant `actors` and `services`?

File: .changeset/fair-steaks-kneel.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+---
+'@xstate/machine-extractor': minor
+---
+
+- Support both actors and guards in machine implementations and if both are present, merges them into a single object
```
Comment (Andarist) [inline .changeset/fair-steaks-kneel.md:5]: This sounds incorrect. `guards` and `actors` probably shouldn't be merged into a single object. Or have you maybe meant `actors` and `services`?

File: .changeset/fair-steaks-kneel.md
Comment (Andarist) [inline .changeset/fair-steaks-kneel.md:1]: This changeset has 2 unrelated things listed here - those should just be 2 separate changesets

File: packages/machine-extractor/src/MachineExtractResult.ts [hunk 5]
```diff
@@ -708,6 +709,46 @@ export class MachineExtractResult {
...
+    ) {
+      return { actions: {}, actors: {}, guards: {}, delays: {} };
+    }
+
+    const out: Record<
+      'actions' | 'actors' | 'guards' | 'delays',
+      Record<string, string>
+    > = { actions: {}, actors: {}, guards: {}, delays: {} } as const;
+
+    const foundKeyToOutKey: Record<string, keyof typeof out> = {
```
Comment (Andarist) [inline packages/machine-extractor/src/MachineExtractResult.ts:725]: nit: IMHO this is a little bit overcomplicated for what it is. A much simpler version would be something like:
```ts
return {
  actors: { ...getThing(this.machineCallResult.options.services), ...getThing(this.machineCallResult.options.actors) },
  actions: getThing(this.machineCallResult.options.actions)
  delays: getThing(this.machineCallResult.options.delays),
  guards: getThing(this.machineCallResult.options.guards)
}
```

File: .changeset/fair-steaks-kneel.md (hunk 4 above)
Comment (farskid) [reply .changeset/fair-steaks-kneel.md:5]: Yeah, it's a typo. It's supposed to be `actors` and `services`.

File: packages/machine-extractor/src/MachineExtractResult.ts (hunk 5 above)
Comment (farskid) [reply packages/machine-extractor/src/MachineExtractResult.ts:725]: Sure, I'll update.

File: .changeset/fair-steaks-kneel.md
Comment (farskid) [reply .changeset/fair-steaks-kneel.md:1]: So 2 separate changesets for a single PR or two separate PRs?

Comment (Andarist) [reply .changeset/fair-steaks-kneel.md:1]: 2 separate changesets work OK

File: .changeset/fair-steaks-kneel.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+'@xstate/machine-extractor': minor
+---
+
+Support both actors and guards in machine implementations and if both are present, merges them into a single object
```
Comment (Andarist) [inline .changeset/fair-steaks-kneel.md]:
```suggestion
Support both `actors` and `services` in machine implementations and if both are present, merges them into a single object
```

**Review by Andarist — APPROVED** [review]

**Pull request merged**